integration_tests/modules_math.py

```diff
@@ -0,0 +1,9 @@
+def factorial(x: i32) -> i64:
+    if x < 0:
+        return 0
+    result: i64
+    result = 1
+    while x > 0:
+        result = result * x
+        x -= 1
+    return result
```

Comment:
With this factorial function (using result as `i64`), I get the following error:
<details>
<summary> Error </summary>
<p>

```
@0 = private unnamed_addr constant [6 x i8] c"%lld\0A\00", align 1

define i32 @main() {
.entry:
  %_lpython_return_variable = alloca i32, align 4
  %answer = alloca i64, align 8
  %x = alloca i32, align 4
  store i32 6, i32* %x, align 4
  %0 = call i64 @__module_modules_math_factorial(i32* %x)
  store i64 %0, i64* %answer, align 4
  %1 = load i64, i64* %answer, align 4
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @0, i32 0, i32 0), i64 %1)
  store i32 0, i32* %_lpython_return_variable, align 4
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return, %.entry
  %2 = load i32, i32* %_lpython_return_variable, align 4
  ret i32 %2
}

define i64 @__module_modules_math_factorial(i32* %x) {
.entry:
  %_lpython_return_variable = alloca i64, align 8
  %result = alloca i64, align 8
  %0 = load i32, i32* %x, align 4
  %1 = icmp slt i32 %0, 0
  br i1 %1, label %then, label %else

then:                                             ; preds = %.entry
  store i32 0, i64* %_lpython_return_variable, align 4
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %ifcont

else:                                             ; preds = %.entry
  br label %ifcont

ifcont:                                           ; preds = %else, %unreachable_after_return
  store i32 1, i64* %result, align 4
  br label %loop.head

loop.head:                                        ; preds = %loop.body, %ifcont
  %2 = load i32, i32* %x, align 4
  %3 = icmp sgt i32 %2, 0
  br i1 %3, label %loop.body, label %loop.end

loop.body:                                        ; preds = %loop.head
  %4 = load i64, i64* %result, align 4
  %5 = load i32, i32* %x, align 4
  %6 = mul i64 %4, i32 %5
  store i64 %6, i64* %result, align 4
  %7 = load i32, i32* %x, align 4
  %8 = sub i32 %7, 1
  store i32 %8, i32* %x, align 4
  br label %loop.head

loop.end:                                         ; preds = %loop.head
  %9 = load i64, i64* %result, align 4
  store i64 %9, i64* %_lpython_return_variable, align 4
  br label %return

unreachable_after_return1:                        ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return1, %loop.end, %then
  %10 = load i64, i64* %_lpython_return_variable, align 4
  ret i64 %10
}

declare void @_lfortran_printf(i8*, ...)
code generation error: asr_to_llvm: module failed verification. Error:
Stored value type does not match pointer operand type!
  store i32 0, i64* %_lpython_return_variable, align 4
 i64Stored value type does not match pointer operand type!
  store i32 1, i64* %result, align 4
 i64Both operands to a binary operator are not of the same type!
  %6 = mul i64 %4, i32 %5
```
</p>
</details>

After, modifying the `factorial` function as below:
<details>
<summary> Code </summary>
<p>

```py
def factorial(x: i32) -> i32:
    if x < 0:
        return 0
    result: i32
    result = 1
    while x > 0:
        result = result * x
        x -= 1
    return result
```
<p>
</details>

<details>
<summary> Error </summary>
<p>

```
Running...
+ integration_tests/expr_01
25
+ python integration_tests/py_expr_01.py
+ integration_tests/expr_02
+ python integration_tests/py_expr_02.py
+ integration_tests/expr_03
120
+ python integration_tests/py_expr_03.py
+ integration_tests/expr_04
21
+ python integration_tests/py_expr_04.py
+ integration_tests/expr_05
-50
+ python integration_tests/py_expr_05.py
+ integration_tests/modules_01
25
OK
+ python integration_tests/py_modules_01.py
+ integration_tests/test_math
720
+ python integration_tests/py_test_math.py
Traceback (most recent call last):
  File "/home/admin-pc/Smitlunagariya/lpython/integration_tests/py_test_math.py", line 5, in <module>
    from modules_math import factorial
  File "/home/admin-pc/Smitlunagariya/lpython/integration_tests/modules_math.py", line 1, in <module>
    def factorial(x: i32) -> i32:
NameError: name 'i32' is not defined
Command 'python integration_tests/py_test_math.py' failed.
```
</p>
</details>

Reply:
It looks like there is some bug related to implicit casting of i32 to i64.

Reply:
Also, the problem I think now seems to be a type mismatch in `llvm` backend. Does it assume int-32 always as a return variable? I mean does it check if the return type needs to be int-64? Sorry if this is a trivial question, I am not much familiar with `llvm`.

Reply:
Can you share the ASR output for this code?

Reply:
Also, is this code generated by `clang` or `asr_to_llvm.cpp`?

Reply:
> Can you share the ASR output for this code?

<details>
<summary> ASR </summary>
<p>

```
(TranslationUnit 
   (SymbolTable 
      1 
      {
         factorial: 
            (ExternalSymbol 
               1 
               factorial 3 factorial 
               modules_math [] 
               factorial 
               Public), 
         main: 
            (Function 
               (SymbolTable 
                  5 
                  {
                     _lpython_return_variable: 
                        (Variable 
                           5 
                           _lpython_return_variable 
                           ReturnVar () () 
                           Default 
                           (Integer 4 []) 
                           Source 
                           Public 
                           Required .false.), 
                     answer: 
                        (Variable 
                           5 
                           answer 
                           Local () () 
                           Default 
                           (Integer 8 []) 
                           Source 
                           Public 
                           Required .false.), 
                     x: 
                        (Variable 
                           5 
                           x 
                           Local () () 
                           Default 
                           (Integer 4 []) 
                           Source 
                           Public 
                           Required .false.)
                  }) 
               main [] [
               (= 
                  (Var 5 x) (ConstantInteger 6 
                  (Integer 4 [])) ()) 
               (= 
                  (Var 5 answer) 
                  (FunctionCall 1 factorial () [
                     (Var 5 x)] [] 
                     (Integer 4 []) () ()) ()) 
               (Print () [
                  (Var 5 answer)]) 
               (= 
                  (Var 5 _lpython_return_variable) (ConstantInteger 0 
                  (Integer 4 [])) ()) 
               (Return)] 
               (Var 5 _lpython_return_variable) 
               Source 
               Public 
               Implementation ()), 
         main_program: 
            (Program 
               (SymbolTable 
                  6 
                  {
                  }) 
               main_program [] []), 
         modules_math: 
            (Module 
               (SymbolTable 
                  3 
                  {
                     factorial: 
                        (Function 
                           (SymbolTable 
                              4 
                              {
                                 _lpython_return_variable: 
                                    (Variable 
                                       4 
                                       _lpython_return_variable 
                                       ReturnVar () () 
                                       Default 
                                       (Integer 4 []) 
                                       Source 
                                       Public 
                                       Required .false.), 
                                 result: 
                                    (Variable 
                                       4 
                                       result 
                                       Local () () 
                                       Default 
                                       (Integer 4 []) 
                                       Source 
                                       Public 
                                       Required .false.), 
                                 x: 
                                    (Variable 
                                       4 
                                       x 
                                       In () () 
                                       Default 
                                       (Integer 4 []) 
                                       Source 
                                       Public 
                                       Required .false.)
                              }) 
                           factorial [
                           (Var 4 x)] [
                           (If 
                              (Compare 
                                 (Var 4 x) 
                                 Lt (ConstantInteger 0 
                                 (Integer 4 [])) 
                                 (Logical 4 []) () ()) [
                              (= 
                                 (Var 4 _lpython_return_variable) (ConstantInteger 0 
                                 (Integer 4 [])) ()) 
                              (Return)] []) 
                           (= 
                              (Var 4 result) (ConstantInteger 1 
                              (Integer 4 [])) ()) 
                           (WhileLoop 
                              (Compare 
                                 (Var 4 x) 
                                 Gt (ConstantInteger 0 
                                 (Integer 4 [])) 
                                 (Logical 4 []) () ()) [
                              (= 
                                 (Var 4 result) 
                                 (BinOp 
                                    (Var 4 result) 
                                    Mul 
                                    (Var 4 x) 
                                    (Integer 4 []) () ()) ()) 
                              (= 
                                 (Var 4 x) 
                                 (BinOp 
                                    (Var 4 x) 
                                    Sub (ConstantInteger 1 
                                    (Integer 4 [])) 
                                    (Integer 4 []) () ()) ())]) 
                           (= 
                              (Var 4 _lpython_return_variable) 
                              (Var 4 result) ()) 
                           (Return)] 
                           (Var 4 _lpython_return_variable) 
                           Source 
                           Public 
                           Implementation ())
                  }) 
               modules_math [] .false.)
      }) 
   [])
```
<p>
</details>

> Also, is this code generated by clang or asr_to_llvm.cpp?

I think `asr_to_llvm.cpp` (not sure though)

Reply:
So, `asr` is generated by `python_ast_to_asr.cpp` and NOT `ast_to_asr.cpp`, right?

Reply:
Yes

Reply:
Yes, we only use `python_ast_to_asr`. We'll remove `ast_to_asr`, which is remaining from LFortran.

Reply:
> It looks like there is some bug related to implicit casting of i32 to i64.

Fixed it! here: https://github.com/lcompilers/lpython/pull/103/commits/3c6ce758ec593476fc0760f514fa4645e95a2efc